zuul-core/src/main/java/com/netflix/zuul/netty/server/psk/ZuulPskServer.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    @Override
+    public TlsCredentials getCredentials() {
+        return null;
+    }
+
+    @Override
+    protected Vector getProtocolNames() {
+        Vector protocolNames = new Vector();
+        if (supportedApplicationProtocols!=null) {
+            supportedApplicationProtocols.forEach(protocolNames::addElement);
+        }
+        return protocolNames;
+    }
+
+    @Override
+    public void notifyHandshakeBeginning() throws IOException {
+        pskTimings.recordHandshakeStarting();
+        this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(false);
+        // TODO: sunnys - handshake timeouts
+        super.notifyHandshakeBeginning();
+    }
+
+    @Override
+    public void notifyHandshakeComplete() throws IOException {
+        pskTimings.recordHandshakeComplete();
+        this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(true);
+        super.notifyHandshakeComplete();
+        ctx.fireUserEventTriggered(SslHandshakeCompletionEvent.SUCCESS);
+    }
+
+    @Override
+    protected ProtocolVersion[] getSupportedVersions() {
+        return ProtocolVersion.TLSv13.only();
+    }
+
+    @Override
+    protected int[] getSupportedCipherSuites() {
+        return TlsUtils.getSupportedCipherSuites(
+                getCrypto(),
+                TlsPskHandler.SUPPORTED_TLS_PSK_CIPHER_SUITE_MAP.keySet().stream()
+                        .mapToInt(Number::intValue)
+                        .toArray());
+    }
+
+    @Override
+    public ProtocolVersion getServerVersion() throws IOException {
+        return super.getServerVersion();
+    }
+
+    @Override
+    @SneakyThrows
+    // TODO: Ask BC folks to see if getExternalPSK can throw a checked exception
+    public TlsPSKExternal getExternalPSK(Vector clientPskIdentities) {
+        byte[] clientPskIdentity = ((PskIdentity)clientPskIdentities.get(0)).getIdentity();
+        byte[] psk;
+        try{
+            this.ctx.channel().attr(TlsPskHandler.CLIENT_PSK_IDENTITY_ATTRIBUTE_KEY).set(new ClientPSKIdentityInfo(clientPskIdentity));
+            psk = externalTlsPskProvider.provide(clientPskIdentity, this.context.getSecurityParametersHandshake().getClientRandom());
+        }catch (PskCreationFailureException e) {
+            throw switch (e.getTlsAlertMessage()) {
+                case unknown_psk_identity -> new TlsFatalAlert(AlertDescription.unknown_psk_identity, "Unknown or null client PSk identity");
+                case decrypt_error -> new TlsFatalAlert(AlertDescription.decrypt_error, "Invalid or expired client PSk identity");
+            };
+        }
+        TlsSecret pskTlsSecret = getCrypto().createSecret(psk);
+        int prfAlgorithm = getPRFAlgorithm13(getSelectedCipherSuite());
+        return new BasicTlsPSKExternal(clientPskIdentity, pskTlsSecret, prfAlgorithm);
+    }
+
+    @Override
+    public void notifyAlertRaised(short alertLevel, short alertDescription, String message, Throwable cause) {
```

Comment:
There is one other event netty fires that we'd be interested in (if possible): https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslCloseCompletionEvent.java netty will fire this event when the the ssl session is closed. Does bouncy castle trigger any similar alerts or anything when that happens?

Reply:
Justin
In TLSProtocol.java file I see it raises close_notify alert : https://github.com/bcgit/bc-java/blob/main/tls/src/main/java/org/bouncycastle/tls/TlsProtocol.java#L300 which will eventually make it's way to this notifyAlertRaised method. Did you want us to do something specific with that alert?
https://github.com/bcgit/bc-java/blob/main/tls/src/main/java/org/bouncycastle/tls/TlsProtocol.java#L1717

Reply:
I added code to send event SslCloseCompletionEvent on close_notify alert